### PR TITLE
Add Protocolumns

### PR DESCRIPTION
**Pull request by CorgiTaco**
Allows the usage of `.getChunk` from CubeWorldGenRegion

### CODE REVIEW

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/biome/ProtoColumnBiomeContainer.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                        + "type: " + biomeContainer.getClass().getSimpleName());
+                }
+            }
+        }
+        this.containers = containers;
+    }
+
+    protected ProtoColumnBiomeContainer(IdMap<Biome> idMap, LevelHeightAccessor levelHeightAccessor, Biome[] biomes) {
+        super(idMap, levelHeightAccessor, biomes);
+        throw new UnsupportedOperationException("We are Using the wrong constructor!");
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/biome/ProtoColumnBiomeContainer.java:50]: Ideally these messages should tell the user which constructor to use, thus avoiding confusion/unnecessary work.

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/biome/ProtoColumnBiomeContainer.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    public ProtoColumnBiomeContainer(IdMap<Biome> idMap, LevelHeightAccessor levelHeightAccessor, ChunkPos chunkPos, BiomeSource biomeSource, @Nullable int[] is) {
+        super(idMap, levelHeightAccessor, chunkPos, biomeSource, is);
+        throw new UnsupportedOperationException("We are Using the wrong constructor!");
+    }
+
+    @Override public int[] writeBiomes() {
+        throw new UnsupportedOperationException("We should not be saving the data in this container!");
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/biome/ProtoColumnBiomeContainer.java:69]: Explain why we shouldn't do that. Point to the place where saving is supposed to happen.

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+import net.minecraft.world.level.levelgen.Heightmap;
+import net.minecraft.world.level.levelgen.feature.StructureFeature;
+import net.minecraft.world.level.levelgen.structure.StructureStart;
+import net.minecraft.world.level.lighting.LevelLightEngine;
+import net.minecraft.world.level.material.Fluid;
+import net.minecraft.world.level.material.FluidState;
+import org.jetbrains.annotations.Nullable;
+
+//TODO: Implement this properly for mods. Vanilla is fine.
+public class CubeWorldRegionColumn extends ProtoChunk {
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java:46]: "CubicWorld"?

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+    private final int xSectionOffset;
+    private final int zSectionOffset;
+    private final IBigCube mainCube;
+    private final IBigCube[] delegates;
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java:51]: I'd rename this to `cubes`. Delegate isn't meaningful.

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+    @Override public boolean isUnsaved() {
+        return false;
+    }
+
+    @Override public ChunkStatus getStatus() {
+        return this.mainCube.getStatus();
+    }
+
+    @Override public void setStatus(ChunkStatus status) {
+        throw new UnsupportedOperationException("This operation is not supported.");
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java:177]: Why? Explain thyself.

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+    @Override public void addEntity(Entity entity) {
+        getCube(entity.blockPosition()).addEntity(entity);
+    }
+
+    @Override public List<CompoundTag> getEntities() {
+        return super.getEntities();
+    }
+
+    @Override public void setBiomes(ChunkBiomeContainer biomes) {
+        throw new UnsupportedOperationException("This is not supported.");
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java:157]: Why not?

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+        throw new UnsupportedOperationException("This is not yet implemented.");
+    }
+
+    @Override public ShortList[] getPackedLights() {
+        throw new UnsupportedOperationException("This is not yet implemented.");
+    }
+
+    @Override public void addLight(short chunkSliceRel, int sectionY) {
+        throw new UnsupportedOperationException("This is not yet implemented.");
+    }
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java:111]: Are any of these three methods required for this PR to be of use?

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+
+    @Override public Heightmap getOrCreateHeightmapUnprimed(Heightmap.Types type) {
+        return this.heightmaps.computeIfAbsent(type, (typex) -> {
+            return new DummyHeightmap(this, typex); //Essentially do nothing here.
+        });
+    }
+
+    @Override public int getHeight(Heightmap.Types heightmapType, int x, int z) {
+        int yStart = cubeToMinBlock(mainCube.getCubePos().getY() + 1);
+        int yEnd = cubeToMinBlock(mainCube.getCubePos().getY());
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java:196]: Why do we ignore the other cubes? Is the main cube the only cube that we care about?

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+
+        IBigCube cube1 = getCube(new BlockPos(x, yStart, z));
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java:198]: Why dont we get this cube using `delegates[delegates.length / 2 + 1]`? Its always the cube above `mainCube`.

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+        if (cube1.getCubeLocalHeight(heightmapType, x, z) >= yStart) {
+            return this.cubeWorldGenRegion.getMinBuildHeight() - 1;
+        }
+        IBigCube cube2 = getCube(new BlockPos(x, yEnd, z));
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java:202]: `yEnd = cubeToMinBlock(mainCube.getCubePos().getY())`. Doesn't this simply return `mainCube`?

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+    private final Map<Heightmap.Types, Heightmap> heightmaps;
+    private final ProtoColumnBiomeContainer protoColumnBiomeContainer;
+
+
+    public CubeWorldRegionColumn(ChunkPos chunkPos, int xSectionOffset, int zSectionOffset, UpgradeData upgradeData, IBigCube[] delegates, CubeWorldGenRegion cubeWorldGenRegion) {
+        super(chunkPos, upgradeData, cubeWorldGenRegion);
+        this.xSectionOffset = xSectionOffset;
+        this.zSectionOffset = zSectionOffset;
+
+        this.mainCube = delegates[delegates.length / 2];
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java:63]: Does this cube have any special purpose or is it simply the one in the middle? In that case `centerCube` would be a more descriptive name.

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+        return mutableBlockPos != null ? mutableBlockPos.immutable() : new BlockPos(chunkPos.getMinBlockX(), this.getHeight(types, chunkPos.getMinBlockX() & 15,
+            chunkPos.getMinBlockZ() & 15), chunkPos.getMinBlockZ() & 15);
+    }
+
+    @Nullable @Override public StructureStart<?> getStartForFeature(StructureFeature<?> structure) {
+        throw new UnsupportedOperationException("This is not yet implemented.");
+    }
+
+    @Override public void setStartForFeature(StructureFeature<?> structure, StructureStart<?> start) {
+        this.getCube(Coords.blockToCube(start.getBoundingBox().minY())); //TODO: Wait on load order
```
Comment (Cyclonit) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubeWorldRegionColumn.java:237]: Load order is done. Is this TODO resolved?

**Review by Cyclonit — COMMENTED** [review]
I didn't notice that this was a draft. Anyways, here we go.

**Comment by CorgiTaco** [comment]
To be done in https://github.com/OpenCubicChunks/CubicChunks2

**Pull request closed without merging**